### ISSUE TITLE
GeneMark.hmm 400-day license.

### ISSUE THREAD

**Issue by fasemoreakinyemi** [ablab/quast#97]
GeneMark.hmm 400-day license.
Your 400-day license period has ended. We hope that you found this
software useful. If you would like to renew this license,
please contact GeneProbe, Inc. at custserv@genepro.com

GeneMarkS: error on last system call, error code 256
Abort program!!!

**Comment by alexeigurevich** [comment]
Hello!
We provide an initial GeneMark license file inside the Quast package and users can freely update it by themselves when the license is expired (after 400 days). We will update the license file in the next Quast release (v.5.1).

So, now you need to get a license from GeneProbe yourself. It is pretty straightforward:
1) Go to [http://exon.gatech.edu/GeneMark/license_download.cgi](http://exon.gatech.edu/GeneMark/license_download.cgi), fill the requested fields are read the license text. Note: you can select _any_ tool and platform actually, e.g. GeneMark-ES / ET v.4.33  and LINUX64.
2) After pressing "I agree ..." button you will be redirected to a download page. You will need to download the key only (either 32bit or 64bit depending on your platform, I think now everyone has 64bit), the software is in Quast package already.
3) The key is in gzip format, you should unpack it (with gunzip) and move to `~/.gm_key`  (the name should be exactly like this, with a dot in the beginning).

After that just rerun Quast and it should run GeneMark software properly.

Let me know if something is unclear or not working!

Disclaimer: we have an agreement with GeneProbe Inc. on distribution of their software inside our package for academic use (see our and their License for details)

**Comment by amizeranschi** [comment]
Hi,

I'm having trouble getting GeneMark-ES to work in QUAST (v. 5.0.2). I registered on the GeneMark-ES website, and got the key, gunzipped and copied to `~/.gm_key` (I tried both the 64bit and 32bit keys), and yet QUAST still won't run with `--gene-finding`:

```
WARNING: GeneMark tool can't be started because of license limitations!
```

If it makes any difference, I'm also running QUAST with the `--fungus` option, as my genome is fungal.

**Comment by splaisan** [comment]
Hi Alexei,

I am also having the error although I installed the sourceforge build and even copied my keys to the package location where I found their counterparts.
In summary I have:
fresh versions of the 

```
.gmhmmp2_key
.gm_key
.gm_key_64
.gms2_key_64
```
keys copied to my home folder

+ the .gm_key_64 file copied to both quast subfolders where keys are found to replace the sourceforge copies:

```
./build/lib/quast_libs/genemark/gm_keys/gm_key_32
=> ./build/lib/quast_libs/genemark/gm_keys/gm_key_64
./quast_libs/genemark/gm_keys/gm_key_32
=> ./quast_libs/genemark/gm_keys/gm_key_64
```

This is clearly not yet enough to get the key seen when quast runs

Any idea?

> quast.py -t ${thr} -r ${ref} -g ${genes} --nanopore reads/${reads} --nanopore reads/${reads2} --conserved-genes-finding --gene-finding --rna-finding --circos -o ${pfx}_${pfxr}_quast assemblies/*.fasta


> 2019-11-19 10:39:17
> Running GeneMarkS...
>   1  ABD384_ABH815.contigs
>   2  ABD384_ABH815.medaka_consensus
>   3  ABD384.contigs
>   4  ABD384.medaka_consensus
>   5  ABH815.contigs
>   6  ABH815.medaka_consensus
> 
> WARNING: License period for GeneMark has ended! 
> To update license, please visit http://exon.gatech.edu/GeneMark/license_download.cgi page and fill in the form.
> You should choose GeneMarkS tool and your operating system (note that GeneMark is free for non-commercial use).
> Download the license key and replace your ~/.gm_key with the updated version. After that you can restart QUAST.
> 
> 
> NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.
>

**Comment by alexeigurevich** [comment]
Hi,
The correct and up-to-date key for you 64/32 bit machine should be present in `~/.gm_key` file, that is in your home directory and named exactly like this, with a dot and without specification of the number of bits. 
Most likely you have the fresh keys in `<...>/quast_libs/genemark/gm_keys/` dirs as you wrote but an old one in `~/.gm_key`. Quast normally tries to substitute the key in `~/` with the corresponding one in `<...>/quast_libs/genemark/gm_keys/` but it checks for the file modification/creation date and this could be elder than your `~/.gm_key` for some reason.

To sum up, please try manually replace the `~/.gm_key` file with your fresh and correct key and restart Quast. Let me know if it still doesn't work.

**Comment by splaisan** [comment]
You are right, I did the following mistakes
copied the gm_key from the gmsuite to ~/.gm_key
and
the gm_key_64 from the download to ~/.gm_key_64

the first one is an old key provided with the package and is not valid anymore

For other dummies like me, all you need to do is copy **and rename** the downloaded and unzipped fresh key **gm_key_64** to  **~/.gm_key**
the other key from the gmsuite is _not to be used_

Thanks Sergei :-)

**Comment by asantat** [comment]
I have an error running genemark.
I have downloaded and unzipped a fresh version of GenemarkS license file into ~/.gm_key. The Quast of v.5.1.0rc1 was downloaded as tar.gz archive with genemark supplied inside. But an error still occures.
While I use a license file supplied within by coping it to ~/.gm_key nothing is changed.

**Comment by alexeigurevich** [comment]
Hi! I've tried to reproduce the issue on both macOS and Linux 64bit just now and everything is working properly. I suppose you did something a bit wrong. Since it is a common issue, I will put a detailed instruction with screenshots below.

**Step 1**. In your browser, go to http://exon.gatech.edu/GeneMark/license_download.cgi

**Step 2**. Select the proper tool and your operating system. If not sure about the tool select the first one, I believe license files are compatible with all of them. Quast actually uses GeneMarkS (default, for prokaryotes), GeneMark-ES (for eukaryotes, i.e. when `--eukaryote` or `--fungus` is specified), and MetaGeneMark (in the metaQuast mode via `./metaquast.py`).
![image](https://user-images.githubusercontent.com/620893/114566575-190b1800-9c7b-11eb-9e2f-43da908a98da.png)

**Step 3**. Fill in the online form, read the license, and sign it if you agree with the terms by pressing the corresponding button. **Note**: GeneMark is free for academic use but commercial use is prohibited, some other restrictions also apply.
![image](https://user-images.githubusercontent.com/620893/114567685-2543a500-9c7c-11eb-9b63-c7c93808630f.png)

**Step 4**. Press on "64 bit" to download the key for the 64-bit system (or if you are absolutely sure that your system is 32 bit, download it). You can also just right-click and copy the link address to download the key in the command line with `wget` (see below). **Note**: 
![image](https://user-images.githubusercontent.com/620893/114568193-9a16df00-9c7c-11eb-82f9-7c545969b13c.png)

**Step 5**. Switch to the terminal and execute the following commands. **Note**: you should use YOUR personal temporary link with `wget` below. 
```
wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_uknWu/gm_key_64.gz  # Note: if you downloaded the key via browser, you may need to execute something like "mv ~/Downloads/gm_key_64.gz ." instead
gunzip gm_key_64.gz  # after this, you should have "gm_key_64" file without extension and readable via e.g. less
mv gm_key_64 ~/.gm_key  # mind the dot in the destination file path. I recommend copy-pasting this command from here
```

**Step 6**. Restart Quast to ensure that gene prediction is working again.
`./quast.py test_data/contigs_1.fasta --gene-find`

**Comment by asantat** [comment]
Thank you very mach for a great explanation.
The problem was in grid system properties. The script has no access to any files outside the allocated space. I just had created a symlink to gm_key_64 inside my work directory with name "~/.gm_key" and the grid system has seen it. So i will rerun all my analysis for correct results to obtain.
